### PR TITLE
Fixed pyright issues in engines directory (#572)

### PR DESCRIPTION
**Pull request by aodhanroche**

### CODE REVIEW

File: griptape/artifacts/list_artifact.py [hunk 1]
```diff
@@ -1,11 +1,11 @@
-from typing import Optional
+from typing import Optional, Sequence
 from attr import field, define
 from griptape.artifacts import BaseArtifact
 
 
 @define
 class ListArtifact(BaseArtifact):
-    value: list[BaseArtifact] = field(factory=list)
+    value: Sequence[BaseArtifact] = field(factory=list)
```
Comment (SavagePencil) [inline griptape/artifacts/list_artifact.py:8]: why `Sequence`? Isn't `list` actually more explicit?

File: griptape/tools/aws_iam_client/tool.py [hunk 2]
```diff
@@ -1,20 +1,15 @@
...
-from typing import TYPE_CHECKING, Optional
 from schema import Schema, Literal
 from attr import define, field, Factory
 from griptape.artifacts import TextArtifact, ErrorArtifact, ListArtifact
 from griptape.utils.decorators import activity
 from griptape.tools import BaseAwsClient
-
-if TYPE_CHECKING:
-    import boto3
+from mypy_boto3_iam import Client
```
Comment (SavagePencil) [inline griptape/tools/aws_iam_client/tool.py:7]: 1. Are we assuming `TYPE_CHECKING` is always True?
2. Is anything different between `mypy_boto3_iam` and the "real" `boto3`?

File: griptape/tools/aws_s3_client/tool.py [hunk 3]
```diff
@@ -182,20 +178,20 @@ def upload_content_to_s3(self, params: dict) -> ErrorArtifact | InfoArtifact:
...
+                artifacts.append(BlobArtifact(content))
 
             except Exception as e:
                 return ErrorArtifact(f"error downloading objects from bucket: {e}")
 
-        return artifact
+        return ListArtifact(artifacts)
 
-    def _upload_object(self, bucket_name: str, object_name: str, value: any) -> None:
+    def _upload_object(self, bucket_name: str, object_name: str, value: Any) -> None:
```
Comment (SavagePencil) [inline griptape/tools/aws_s3_client/tool.py:194]: for my own edification: what's the difference between `any` and `Any`?

**Comment by aodhanroche** [comment]
Sorry, Did this backwards

File: griptape/tools/base_aws_client.py [hunk 4]
```diff
@@ -1,5 +1,5 @@
 from __future__ import annotations
-from typing import TYPE_CHECKING, Optional
+from typing import TYPE_CHECKING
```
Comment (SavagePencil) [inline griptape/tools/base_aws_client.py:2]: this file continues to use the `if TYPE_CHECKING` `import boto3` pattern, but elsewhere we've stopped doing that in favor of importing a `mypy_boto3`. Why not in this file?

File: griptape/tools/base_tool.py [hunk 5]
```diff
@@ -53,6 +53,8 @@ def validate_output_memory(self, _, output_memory: dict[str, Optional[list[TaskM
             for activity_name, memory_list in output_memory.items():
                 if not self.find_activity(activity_name):
                     raise ValueError(f"activity {activity_name} doesn't exist")
+                if memory_list is None:
+                    raise ValueError(f"memory list for activity '{activity_name}' can't be None")
```
Comment (SavagePencil) [inline griptape/tools/base_tool.py:57]: but the function signature says it's `Optional`, so why can't it be `None`?

File: griptape/tools/base_tool.py [hunk 6]
```diff
@@ -92,7 +94,9 @@ def schema(self) -> dict:
                 {
                     Literal("name"): self.name,
                     Literal("path", description=self.activity_description(activity)): self.activity_name(activity),
-                    Literal("input"): {"values": activity.config["schema"]} if self.activity_schema(activity) else {},
+                    Literal("input"): {"values": getattr(activity, "config")["schema"]}
+                    if self.activity_schema(activity)
+                    else {},
```
Comment (SavagePencil) [inline griptape/tools/base_tool.py:99]: this is hurting my brain can we explain this

File: griptape/tools/base_tool.py [hunk 7]
```diff
@@ -156,7 +161,7 @@ def tool_dir(self):
 
         return os.path.dirname(os.path.abspath(class_file))
 
-    def install_dependencies(self, env: dict[str, Optional[str]] = None) -> None:
+    def install_dependencies(self, env: Optional[dict[str, str]] = None) -> None:
```
Comment (SavagePencil) [inline griptape/tools/base_tool.py:164]: This feels nuanced. In the previous line, it was a dictionary of valid keys and optional second params, but here we're saying the dict is optional. Can we explain?

File: griptape/tools/calculator/tool.py [hunk 8]
```diff
@@ -20,7 +20,7 @@ class Calculator(BaseTool):
         }
     )
     def calculate(self, params: dict) -> BaseArtifact:
-        import numexpr
+        import numexpr  # type: ignore
```
Comment (SavagePencil) [inline griptape/tools/calculator/tool.py:23]: why dat

File: griptape/tools/computer/tool.py [hunk 9]
```diff
@@ -106,18 +107,21 @@ def execute_command_in_container(self, command: str) -> BaseArtifact:
                 detach=True,
             )
 
-            container.wait()
+            if isinstance(container, Container):
+                container.wait()
```
Comment (SavagePencil) [inline griptape/tools/computer/tool.py:111]: Should we do something different if it's NOT a `Container`?

File: griptape/tools/computer/tool.py [hunk 10]
```diff
@@ -106,18 +107,21 @@ def execute_command_in_container(self, command: str) -> BaseArtifact:
...
 
-            if stderr:
-                return ErrorArtifact(stderr)
+                if stderr:
+                    return ErrorArtifact(stderr)
+                else:
+                    return TextArtifact(stdout)
             else:
-                return TextArtifact(stdout)
+                return ErrorArtifact("error running container")
```
Comment (SavagePencil) [inline griptape/tools/computer/tool.py:124]: do we lose the contents of `stdout` this way? Should we f-string it?

File: griptape/tools/computer/tool.py [hunk 11]
```diff
@@ -161,9 +165,10 @@ def container_name(self, tool: BaseTool) -> str:
     def remove_existing_container(self, name: str) -> None:
         try:
             existing_container = self.docker_client.containers.get(name)
-            existing_container.remove(force=True)
+            if isinstance(existing_container, Container):
```
Comment (SavagePencil) [inline griptape/tools/computer/tool.py:168]: as before, should we handle the case when it's not a `Container`?

File: griptape/tools/computer/tool.py [hunk 12]
```diff
@@ -174,14 +179,15 @@ def build_image(self, tool: BaseTool) -> None:
 
             image = self.docker_client.images.build(path=temp_dir, tag=self.image_name(tool), rm=True, forcerm=True)
 
-            response = [line for line in image]
+            if isinstance(image, tuple):
```
Comment (SavagePencil) [inline griptape/tools/computer/tool.py:182]: what if it's not that thing

File: griptape/tools/google_drive/tool.py [hunk 13]
```diff
@@ -32,7 +33,7 @@ class GoogleDriveClient(BaseGoogleClient):
             "description": "Can be used to list files in a specific Google Drive folder.",
             "schema": Schema(
                 {
-                    Optional(
+                    schema.Optional(
```
Comment (SavagePencil) [inline griptape/tools/google_drive/tool.py:36]: how is schema's Optional different?

File: griptape/tools/web_scraper/tool.py [hunk 14]
```diff
@@ -20,12 +20,13 @@ class WebScraper(BaseTool):
...
-        if isinstance(page, ErrorArtifact):
-            return page
-        else:
-            return ListArtifact(TextLoader().text_to_artifacts(page.get("text")))
+        try:
+            page = WebLoader().extract_page(url, self.include_links)
+
+            return ListArtifact(TextLoader().text_to_artifacts(page["text"]))
+        except Exception as e:
+            return ErrorArtifact("Error getting page content: " + str(e))
```
Comment (SavagePencil) [inline griptape/tools/web_scraper/tool.py:29]: Any reason we're concatenating vs. f-string?

File: griptape/tools/web_scraper/tool.py [hunk 15]
```diff
@@ -35,9 +36,10 @@ def get_content(self, params: dict) -> ListArtifact | ErrorArtifact:
...
-        if isinstance(page, ErrorArtifact):
-            return page
-        else:
-            return TextArtifact(page.get("author"))
+        try:
+            page = WebLoader().extract_page(url, self.include_links)
+
+            return TextArtifact(page["author"])
+        except Exception as e:
+            return ErrorArtifact("Error getting page author: " + str(e))
```
Comment (SavagePencil) [inline griptape/tools/web_scraper/tool.py:45]: Any reason we're concatenating vs. f-string?

**Review by SavagePencil — CHANGES_REQUESTED** [review]
Lotta questions, a bunch of them are probably easily answered and can be resolved.

**Pull request closed without merging**